# Disallows usage of Clarity Angular badges (no-clr-badge)

The use of Clarity Angular badges is discouraged. Use Clarity Core badges instead.

## Rule Details

Examples of **incorrect** code for this rule:

```html
<span class="badge">1</span>
```

Examples of **correct** code for this rule:

```html
<cds-badge>1</cds-badge>
```
