# Disallows usage of Clarity Angular buttons (no-clr-button)

The use of Clarity Angular buttons is discouraged. Use Clarity Core buttons instead.

## Rule Details

Examples of **incorrect** code for this rule:

```html
<button class="btn btn-primary">Open</button>
```

Examples of **correct** code for this rule:

```html
<cds-button>Open</cds-button>
```
