// clarity-angular/wizard/wizard-step.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var core_1 = require("@angular/core");
var tab_link_1 = require("../tabs/tab-link");
var wizard_1 = require("./wizard");
var WizardStep = (function (_super) {
    __extends(WizardStep, _super);
    function WizardStep(wizard, elementRef) {
        _super.call(this, wizard);
        this.wizard = wizard;
        this.elementRef = elementRef;
        // is the section completed
        this.isCompleted = false;
        // input variable, optional, to set if this tab is skipped
        this.isSkipped = false;
    }
    WizardStep.prototype.onClick = function () {
        // IE 10 fix to prevent click of disabled tab
        if (!this.active && this.isCompleted) {
            _super.prototype.onClick.call(this);
        }
        return false;
    };
    WizardStep.prototype.ngOnInit = function () {
        // The nav text will be the title of the main section.
        this.title = this.elementRef.nativeElement.innerText.trim();
    };
    __decorate([
        core_1.Input("clrWizardStepIsSkipped"), 
        __metadata('design:type', Boolean)
    ], WizardStep.prototype, "isSkipped", void 0);
    __decorate([
        core_1.Input("clrWizardStepId"), 
        __metadata('design:type', String)
    ], WizardStep.prototype, "id", void 0);
    WizardStep = __decorate([
        core_1.Component({
            selector: "clr-wizard-step",
            template: "\n      <a href=\"#\" role=\"tab\" (click)=\"onClick()\" class=\"nav-link\">\n          <ng-content></ng-content>\n      </a>\n    ",
            host: {
                "[id]": "id",
                "[attr.aria-selected]": "active",
                "[attr.aria-controls]": "ariaControls",
                "role": "presentation",
                "[class.clr-nav-link]": "true",
                "[class.active]": "active",
                "[class.disabled]": "!active && !isCompleted",
                "[class.complete]": "isCompleted",
                "[class.skipped]": "isSkipped"
            }
        }),
        __param(0, core_1.Inject(core_1.forwardRef(function () { return wizard_1.Wizard; }))), 
        __metadata('design:paramtypes', [wizard_1.Wizard, core_1.ElementRef])
    ], WizardStep);
    return WizardStep;
}(tab_link_1.TabLink));
exports.WizardStep = WizardStep;
